toolchain/lower/file_context.cpp

```diff
@@ -373,10 +373,18 @@ auto FileContext::BuildDISubprogram(const SemIR::Function& function,
   auto loc = converter_.ConvertLoc(
       function.definition_id,
       [](DiagnosticLoc, const Internal::DiagnosticBase<>&) {});
+  llvm::StringRef name;
+  if (function.name_id.is_valid()) {
+    name = *sem_ir().names().GetAsStringIfIdentifier(function.name_id);
```

Comment:
BuildFunctionDecl does similar, had you considered refactoring to use the same name in both places?

Reply:
Like refactor out a helper function to call from both places, or pass the name, retrieved in `BuildFunctionDecl` along through the decl to the def building somehow and then down into `BuildDISubprogram`?

It seemed short enough it wasn't necessarily worth a helper function - but I'm certainly open to it, if that's what you had in mind?

Sharing the value from `BuildFunctionDecl` into the def building code and into the BuildDISubprogram seemed a bit more difficult (because the decl might be built some time before/pretty distant from the definition, I assumed?)

Reply:
Sorry, I think I misunderstood exactly what you were trying to do with this name. I'm realizing two things:

1) You're trying to _avoid_ using a mangled name here, right? Why does this compare with `linkage_name`, versus unconditionally setting the value? Maybe adding comments might be helpful?

2) Have you looked at GetFormatted instead of GetAsStringIfIdentifier?

    - In particular, can you add a test with `fn r#self() {}`? I think we haven't really tested with raw identifiers yet, I'm wondering if it's going to show here.

Reply:
1. Added comment in 0eaa1ecddea8f51aaf50b72dd7c1b8c11bb5a771 - we only want the linkage name if it differs from the source name.

2. Thanks for the offline discussion - adjusted the existing name query/checking to be a bit clearer in 4c9021cbc8f9f50790d72f04aedc25aa70e7291e and added test coverage for raw identifiers in functions (which seems to come out naturally from the identifier handling - so isn't dependent on this patch or other changes) in #4257

Reply:
> 1. Added comment in [0eaa1ec](https://github.com/carbon-language/carbon-lang/commit/0eaa1ecddea8f51aaf50b72dd7c1b8c11bb5a771) - we only want the linkage name if it differs from the source name.

Per discussion, let's get rid of the conditional since it's mainly going to affect "extern C"-style things, which are atypical in Carbon and probably okay to store a name for. For a while, we'll just have name mangling and nothing will match.

Reply:
done